agent/exec/containerd/adapter.go

```diff
@@ -0,0 +1,424 @@
+		return err
+	}
+	return nil
+}
+
+func (c *containerAdapter) create(ctx context.Context) error {
+	mpath := filepath.Join(c.contentStore, "manifests", "docker.io", c.container.Image)
+	mbytes, err := ioutil.ReadFile(mpath)
+	if err != nil {
+		return err
+	}
+
+	var mver manifest.Versioned
+	if err := json.Unmarshal(mbytes, &mver); err != nil {
+		return err
+	}
+
+	rootfs := filepath.Join(c.bundle, "rootfs")
+	stdin := filepath.Join(c.bundle, "stdin")
+	stdout := filepath.Join(c.bundle, "stdout")
+	stderr := filepath.Join(c.bundle, "stderr")
+
+	if err := os.MkdirAll(rootfs, 0755); err != nil {
+		return err
+	}
+
+	// Need github.com/docker/docker/oci.DefaultSpec()
+	spec := specs.Spec{
+		Version: "1.0.0-rc2-dev",
+		Root: specs.Root{
+			Path: "rootfs",
+		},
+		Mounts: []specs.Mount{
+			{
+				Destination: "/proc",
+				Type:        "proc",
+				Source:      "proc",
+				Options:     []string{"nosuid", "noexec", "nodev"},
+			},
+			{
+				Destination: "/dev",
+				Type:        "tmpfs",
+				Source:      "tmpfs",
+				Options:     []string{"nosuid", "strictatime", "mode=755"},
+			},
+			{
+				Destination: "/dev/pts",
+				Type:        "devpts",
+				Source:      "devpts",
+				Options:     []string{"nosuid", "noexec", "newinstance", "ptmxmode=0666", "mode=0620", "gid=5"},
+			},
+			{
+				Destination: "/sys",
+				Type:        "sysfs",
+				Source:      "sysfs",
+				Options:     []string{"nosuid", "noexec", "nodev", "ro"},
+			},
+			{
+				Destination: "/sys/fs/cgroup",
+				Type:        "cgroup",
+				Source:      "cgroup",
+				Options:     []string{"ro", "nosuid", "noexec", "nodev"},
+			},
+			{
+				Destination: "/dev/mqueue",
+				Type:        "mqueue",
+				Source:      "mqueue",
+				Options:     []string{"nosuid", "noexec", "nodev"},
+			},
+		},
+		Process: specs.Process{
+			Cwd: "/",
+			Capabilities: []string{
+				"CAP_CHOWN",
+				"CAP_DAC_OVERRIDE",
+				"CAP_FSETID",
+				"CAP_FOWNER",
+				"CAP_MKNOD",
+				"CAP_NET_RAW",
+				"CAP_SETGID",
+				"CAP_SETUID",
+				"CAP_SETFCAP",
+				"CAP_SETPCAP",
+				"CAP_NET_BIND_SERVICE",
+				"CAP_SYS_CHROOT",
+				"CAP_KILL",
+				"CAP_AUDIT_WRITE",
+			},
+		},
+		Linux: &specs.Linux{
+			Namespaces: []specs.LinuxNamespace{
+				{Type: "mount"},
+				{Type: "network"},
+				{Type: "uts"},
+				{Type: "pid"},
+				{Type: "ipc"},
+			},
+		},
+	}
+
+	// XXX need some proper types from somewhere. github.com/docker/docker/image perhaps.
+	type CConfig struct {
+		Env        []string
+		Cmd        []string
+		Entrypoint []string
+		Volumes    map[string]struct{}
+		WorkingDir string
+	}
+	type Config struct {
+		Architecture string  `json:"architecture"`
+		Os           string  `json:"os"`
+		Config       CConfig `json:"config"`
+	}
+	var config Config
+
+	switch mver.SchemaVersion {
+	case 1:
+		var manifest schema1.Manifest
+		if err := json.Unmarshal(mbytes, &manifest); err != nil {
+			return err
+		}
+
+		// No idea if this is correct...
```

Comment:
Looks right. Seems kind of unfortunate for swarmkit to know anything about manifests though - this seems like something that belongs at a different layer of the stack. Is the plan to move this into containerd once the relevant APIs exist?